File: 백준/Gold/3190. 뱀/뱀.py
import sys
from collections import deque
input=sys.stdin.readline

# 오른쪽을 보고 있음
# 벽 또는 자기 자신과 만나면 게임 끝
# 몸을 늘려 머리를 다음칸에

def find(N,x,y,d,exist):
  if d=='T':
    for i in range(x,-1,-1):
      if l[i][y]==-1:
        break
      if dir and time[0]==int(dir[0][0]):
        if dir[0][1]=='L':
          dir.popleft()
          find(N,i,y,'L',exist)
        else:
          dir.popleft()
          find(N,i,y,'R',exist)
        return

      time[0]+=1
      if l[i][y]==0:
        [prevX,prevY]=exist.popleft()
        l[prevX][prevY]=0
      exist.append([i,y])
      l[i][y]=-1

  elif d=='D':
    for i in range(x,N):
      if l[i][y]==-1:
        break
      if dir and time[0]==int(dir[0][0]):
        if dir[0][1]=='L':
          dir.popleft()
          find(N,i,y,'R',exist)
        else:
          dir.popleft()
          find(N,i,y,'L',exist)
        return

      time[0]+=1
      if l[i][y]==0:
        [prevX,prevY]=exist.popleft()
        l[prevX][prevY]=0
      exist.append([i,y])
      l[i][y]=-1

  elif d=='L':
    for i in range(y,-1,-1):
      if l[x][i]==-1:
        break
      if dir and time[0]==int(dir[0][0]):
        if dir[0][1]=='L':
          dir.popleft()
          find(N,x,i,'D',exist)
        else:
          dir.popleft()
          find(N,x,i,'T',exist)
        return

      time[0]+=1
      if l[x][i]==0:
        [prevX,prevY]=exist.popleft()
        l[prevX][prevY]=0
      exist.append([x,i])
      l[x][i]=-1

  else: # R
    for i in range(y,N):
      if l[x][i]==-1:
        break
      if dir and time[0]==int(dir[0][0]):
        if dir[0][1]=='L':
          dir.popleft()
          find(N,x,i,'T',exist)
        else:
          dir.popleft()
          find(N,x,i,'D',exist)
        return

      time[0]+=1
      if l[x][i]==0:
        [prevX,prevY]=exist.popleft()
        l[prevX][prevY]=0
      exist.append([x,i])
      l[x][i]=-1

  return

N=int(input())
K=int(input())
l=[[0 for i in range(N)]for j in range(N)]

for i in range(K):
  a,b=map(int,input().split())
  l[a-1][b-1]=1

L=int(input()) # 방향 전환 횟수
dir=deque([])
for i in range(L):
  dir.append(list(input().split()))

time=[0]
find(N,0,0,'R',deque([[0,0]]))
print(time[0])

File: 백준/Gold/3190. 뱀/뱀.js
const readline=require('readline')
const rl=readline.createInterface({
  input:process.stdin,
  output:process.stdout
})

const input=[]
let iter=0
rl.on('line',(line)=>{
  input.push(line)
}).on('close',()=>{
  const N=Number(input[iter++])
  const K=Number(input[iter++])

  // 0은 빈칸, 1은 뱀, 2는 사과
  const l=[...new Array(N)].map(item=>[...new Array(N)].map(item=>0))
  l[0][0]=1
  const snail=[[0,0]]
  let headx=0
  let heady=0
  let dir=1 // 위 오 아 왼

  for(let i=0;i<K;i++){
    const [x,y]=input[iter++].split(' ').map(item=>Number(item)-1)
    l[x][y]=2
  }

  let answer=1
  const L=Number(input[iter++])
  let prev=0
  for(let i=0;i<L;i++){
    const [X,C]=input[iter++].split(' ')
    const x=Number(X)-prev
    prev=Number(X)


    for(let j=0;j<x;j++){
      let newx=headx
      let newy=heady
      if (dir===0){
        newx=headx-1
      } else if (dir===1){
        newy=heady+1
      } else if (dir===2){
        newx=headx+1
      } else {
        newy=heady-1
      }
      headx=newx
      heady=newy

      if (0<=newx && newx<N && 0<=newy && newy<N){
        if (l[newx][newy]===1){
          console.log(answer)
          process.exit()
        }
        if (l[newx][newy]===0){
          const [popx,popy]=snail.shift()
          l[popx][popy]=0
        }
        l[newx][newy]=1
        snail.push([newx,newy])
      } else {
        console.log(answer)
        process.exit()
      }

      answer+=1
    }

    if (C==='L') dir=(dir+4-1)%4
    else dir=(dir+1)%4
  }

  while (true){
    let newx=headx
    let newy=heady
    if (dir===0){
      newx=headx-1
    } else if (dir===1){
      newy=heady+1
    } else if (dir===2){
      newx=headx+1
    } else {
      newy=heady-1
    }
    headx=newx
    heady=newy

    if (0<=newx && newx<N && 0<=newy && newy<N){
      if (l[newx][newy]===1){
        console.log(answer)
        process.exit()
      }
      if (l[newx][newy]===0){
        const [popx,popy]=snail.shift()
        l[popx][popy]=0
      }
      l[newx][newy]=1
      snail.push([newx,newy])
    } else {
      console.log(answer)
      process.exit()
    }

    answer+=1
  }
  process.exit()
})
